refiner-app/src/socket.rs

```diff
@@ -0,0 +1,295 @@
+use std::path::Path;
+
+use actix_rt::net::{UnixListener, UnixStream};
+use actix_rt::Arbiter;
+use aurora_standalone_engine::{
+    gas::{estimate_gas, EthCallRequest},
+    tracing::lib::{trace_transaction, DebugTraceTransactionRequest},
+};
+use engine_standalone_storage::Storage;
+use engine_standalone_tracing::types::call_tracer::SerializableCallFrame;
+use serde_json::json;
+use tokio::io::{AsyncReadExt, AsyncWriteExt, Interest};
+
+type SharedStorage = std::sync::Arc<tokio::sync::RwLock<Storage>>;
+
+pub async fn start_socket_server(storage: SharedStorage, path: &Path) {
+    let sock = UnixListener::bind(path).unwrap();
+    let arbiter = Arbiter::new();
+
+    loop {
+        if let Ok((mut stream, _)) = sock.accept().await {
+            let storage = storage.clone();
+            arbiter.spawn(async move {
+                handle_conn(storage, &mut stream).await;
+            });
+        }
+    }
+
+    // arbiter.stop();
+    // arbiter.join().unwrap();
+}
+
+async fn handle_conn(storage: SharedStorage, stream: &mut UnixStream) {
+    match stream.ready(Interest::READABLE | Interest::WRITABLE).await {
+        Ok(r) if r.is_writable() && r.is_readable() => (),
+        r => {
+            let _ = stream.shutdown().await;
+            eprintln!("faulty stream: {:?}", r);
+            return;
+        }
+    };
+    let mut data = vec![0; 1024];
+    loop {
+        match stream.read(&mut data).await {
```

Comment:
I don't think this code quite works the way you want it to. I think if the message doesn't show up all at once (which can happen with async communication) then this code will fail. I think the client should prefix all messages with a 4-byte length. Then you always know exactly how many bytes need to be read for there to be a complete message (always start by reading 4 bytes, then read as many bytes as that number tells you to). You would have a function something like this:

```rust
use tokio::io::{self, AsyncRead, AsyncReadExt};

pub async fn wrapped_read<R: AsyncRead + Unpin>(reader: &mut R) -> io::Result<Vec<u8>> {
    let payload_len = reader.read_u32_le().await? as usize;
    let mut payload = vec![0; payload_len];
    reader.read_exact(&mut payload).await?;
    Ok(payload)
}
```

Note the `reader` that you pass to this function can be anything implementing the `AsyncRead`, but in your case you would always be passing a `UnixStream`, so you don't need to use the type generics if you don't want to.

Similarly, you probably want to respond to the client with the same length-prefix format:

```rust
pub async fn wrapped_write<W: AsyncWrite + Unpin>(
    writer: &mut W,
    payload: &[u8],
) -> io::Result<()> {
    let payload_len = payload.len();
    writer.write_u32_le(payload_len as u32).await?;
    writer.write_all(payload).await
}
```

Reply:
Thank you. I wasn't aware of the exact semantics and it's definitely better to have the message boundaries defined, even if the previous version would have somehow worked.
Added in 5bf322e